mypy/tvar_scope.py

```diff
@@ -0,0 +1,82 @@
+from typing import Optional, Dict, Union
+from mypy.types import TypeVarDef, TypeVarId
+from mypy.nodes import TypeVarExpr, SymbolTableNode
+
+
+class TypeVarScope:
+    """Scope that holds bindings for type variables. Node fullname -> TypeVarDef."""
+
+    def __init__(self,
+                 parent: Optional['TypeVarScope'] = None,
+                 is_class_scope: bool = False,
+                 prohibited: Optional['TypeVarScope'] = None) -> None:
+        """Initializer for TypeVarScope
+
+        Parameters:
+        parent: the outer scope for this scope
+        is_class_scope: True if this represents a generic class
+        prohibited: Type variables that aren't strictly in scope exactly,
+                    but can't be bound because they're part of an outer class's scope.
+        """
+        self.scope = {}  # type: Dict[str, TypeVarDef]
+        self.parent = parent
+        self.func_id = 0
+        self.class_id = 0
+        self.is_class_scope = is_class_scope
+        self.prohibited = prohibited
+        if parent is not None:
+            self.func_id = parent.func_id
+            self.class_id = parent.class_id
+
+    def get_function_scope(self) -> Optional['TypeVarScope']:
+        """Get the nearest parent that's a function scope, not a class scope"""
+        it = self
+        while it.is_class_scope:
+            it = it.parent
+        return it
+
+    def allow_binding(self, fullname: str) -> bool:
+        if fullname in self.scope:
+            return False
+        elif self.parent and not self.parent.allow_binding(fullname):
+            return False
+        elif self.prohibited and not self.prohibited.allow_binding(fullname):
+            return False
+        return True
+
+    def method_frame(self) -> 'TypeVarScope':
+        """A new scope frame for binding a method"""
+        return TypeVarScope(self, False, None)
+
+    def class_frame(self) -> 'TypeVarScope':
+        """A new scope frame for binding a class. Prohibits *this* class's tvars"""
+        return TypeVarScope(self.get_function_scope(), True, self)
```

Comment:
You are always prohibited from re-binding an already-bound type variable. When you define a class in a function, this doesn't prohibit anything more -- your class type variables are also going to be prohibited by dint of being in your own parent.